### ISSUE TITLE
[Regression] dgl.batch() in 0.5.x is slower than 0.4.x

### ISSUE THREAD

**Issue by BarclayII** [dmlc/dgl#2145]
## 🐛 Bug

Not sure if this is a bug, but here is the result:

## To Reproduce

Benchmarking script:
```python
import dgl
import numpy as np
import time
import itertools

for E, B in itertools.product([1000, 10000, 100000], [10, 100, 1000]):
    gs = [dgl.DGLGraph((np.random.randint(0, 100, E), np.random.randint(0, 100, E)), readonly=True) for _ in range(B)]
    t0 = time.time()
    for _ in range(20):
        g = dgl.batch(gs)
    tt = time.time()
    print(E, B, tt - t0)
```

Result on a g3.16x instance (although the code works on CPU):

|E|B|0.5.x|0.4.x|
|:-:|:-:|:-:|:-:|
| 1000 | 10 | 0.0637 | 0.00382 |
| 1000 | 100 | 0.447 | 0.0276 |
| 1000 | 1000 | 4.44 | 0.353 |
| 10000 | 10 | 0.0843 | 0.0138 |
| 10000 | 100 | 0.790 | 0.194 |
| 10000 | 1000 | 10.0 | 2.85 |
| 100000 | 10 | 0.376 | 0.120 |
| 100000 | 100 | 4.11 | 2.38 |
| 100000 | 1000 | 41.0 | 26.2 |

## Context

Vijay and I noticed a significant slowdown on the training of GATs for graph regression on ZINC in Benchmarking GNNs.  The slowdown goes as much as 100% (depending on the machine).  With `CUDA_LAUNCH_BLOCKING=1`, I further figured out that fetching a batched graph from the dataloader increases from 13ms to 89ms.  The forward propagation also increases from 29ms to 41ms per iteration, and the backward propagation increases from 31ms to 43ms, but that's a separate issue.

**Comment by yzh119** [comment]
After #2155 , the speed comparison between 0.4.x and 0.5.x becomes:
|E|B|0.5.x|0.4.x|
|:-:|:-:|:-:|:-:|
| 1000 | 10 | 0.012624025344848633 | 0.0029299259185791016 |
| 1000 | 100 | 0.09022283554077148 | 0.018996477127075195 |
| 1000 | 1000 | 0.9731013774871826 | 0.26607394218444824 |
| 10000 | 10 | 0.0177004337310791 | 0.007367372512817383 |
| 10000 | 100 | 0.21375083923339844 | 0.06772661209106445 |
| 10000 | 1000 | 2.892416000366211 | 1.7698068618774414 |
| 100000 | 10 | 0.09859418869018555 | 0.06327462196350098 |
| 100000 | 100 | 2.1583847999572754 | 1.626404047012329 |
| 100000 | 1000 | 21.13863205909729 | 16.461536407470703 |

Still not good, especially when batching a large number of small graphs.

**Comment by yzh119** [comment]
The overhead of creating a DGLGraph from GraphIndex is non-trivial, if I only create the graph index, the time becomes:
|E|B|0.5.x|0.4.x|
|:-:|:-:|:-:|:-:|
| 1000 | 10 | 0.004979848861694336 | 0.0029299259185791016 |
| 1000 | 100 | 0.03594255447387695 | 0.018996477127075195 |
| 1000 | 1000 | 0.42719340324401855 | 0.26607394218444824 |
| 10000 | 10 | 0.00970005989074707 | 0.007367372512817383 |
| 10000 | 100 | 0.10654783248901367 | 0.06772661209106445 |
| 10000 | 1000 | 2.68953275680542 | 1.7698068618774414 |
| 100000 | 10 | 0.08985686302185059 | 0.06327462196350098 |
| 100000 | 100 | 2.0495126247406006 | 1.626404047012329 |
| 100000 | 1000 | 21.301225662231445 | 16.461536407470703 |

There is a constant gap which is linear to B, about 0.006s per batch.

**Comment by yzh119** [comment]
The profiling result when E = 1000 and B = 1000:
```
Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
    14                                           @profile
    15                                           def batch(graphs, ndata=ALL, edata=ALL, *, node_attrs=None, edge_attrs=None):
    16                                               r"""Batch a collection of :class:`DGLGraph` s into one graph for more efficient
    17                                               graph computation.
    18                                           
    19                                               Each input graph becomes one disjoint component of the batched graph. The nodes
    20                                               and edges are relabeled to be disjoint segments:
    21                                           
    22                                               =================  =========  =================  ===  =========
    23                                                                  graphs[0]  graphs[1]          ...  graphs[k]
    24                                               =================  =========  =================  ===  =========
    25                                               Original node ID   0 ~ N_0    0 ~ N_1            ...  0 ~ N_k
    26                                               New node ID        0 ~ N_0    N_0+1 ~ N_0+N_1+1  ...  1+\sum_{i=0}^{k-1} N_i ~
    27                                                                                                     1+\sum_{i=0}^k N_i
    28                                               =================  =========  =================  ===  =========
    29                                           
    30                                               Because of this, many of the computations on a batched graph are the same as if
    31                                               performed on each graph individually, but become much more efficient
    32                                               since they can be parallelized easily. This makes ``dgl.batch`` very useful
    33                                               for tasks dealing with many graph samples such as graph classification tasks.
    34                                           
    35                                               For heterograph inputs, they must share the same set of relations (i.e., node types
    36                                               and edge types) and the function will perform batching on each relation one by one.
    37                                               Thus, the result is also a heterograph and has the same set of relations as the inputs.
    38                                           
    39                                               The numbers of nodes and edges of the input graphs are accessible via the
    40                                               :func:`DGLGraph.batch_num_nodes` and :func:`DGLGraph.batch_num_edges` attributes
    41                                               of the resulting graph. For homogeneous graphs, they are 1D integer tensors,
    42                                               with each element being the number of nodes/edges of the corresponding input graph. For
    43                                               heterographs, they are dictionaries of 1D integer tensors, with node
    44                                               type or edge type as the keys.
    45                                           
    46                                               The function supports batching batched graphs. The batch size of the result
    47                                               graph is the sum of the batch sizes of all the input graphs.
    48                                           
    49                                               By default, node/edge features are batched by concatenating the feature tensors
    50                                               of all input graphs. This thus requires features of the same name to have
    51                                               the same data type and feature size. One can pass ``None`` to the ``ndata``
    52                                               or ``edata`` argument to prevent feature batching, or pass a list of strings
    53                                               to specify which features to batch.
    54                                           
    55                                               To unbatch the graph back to a list, use the :func:`dgl.unbatch` function.
    56                                           
    57                                               Parameters
    58                                               ----------
    59                                               graphs : list[DGLGraph]
    60                                                   Input graphs.
    61                                               ndata : list[str], None, optional
    62                                                   Node features to batch.
    63                                               edata : list[str], None, optional
    64                                                   Edge features to batch.
    65                                           
    66                                               Returns
    67                                               -------
    68                                               DGLGraph
    69                                                   Batched graph.
    70                                           
    71                                               Examples
    72                                               --------
    73                                           
    74                                               Batch homogeneous graphs
    75                                           
    76                                               >>> import dgl
    77                                               >>> import torch as th
    78                                               >>> # 4 nodes, 3 edges
    79                                               >>> g1 = dgl.graph((th.tensor([0, 1, 2]), th.tensor([1, 2, 3])))
    80                                               >>> # 3 nodes, 4 edges
    81                                               >>> g2 = dgl.graph((th.tensor([0, 0, 0, 1]), th.tensor([0, 1, 2, 0])))
    82                                               >>> bg = dgl.batch([g1, g2])
    83                                               >>> bg
    84                                               Graph(num_nodes=7, num_edges=7,
    85                                                     ndata_schemes={}
    86                                                     edata_schemes={})
    87                                               >>> bg.batch_size
    88                                               2
    89                                               >>> bg.batch_num_nodes()
    90                                               tensor([4, 3])
    91                                               >>> bg.batch_num_edges()
    92                                               tensor([3, 4])
    93                                               >>> bg.edges()
    94                                               (tensor([0, 1, 2, 4, 4, 4, 5], tensor([1, 2, 3, 4, 5, 6, 4]))
    95                                           
    96                                               Batch batched graphs
    97                                           
    98                                               >>> bbg = dgl.batch([bg, bg])
    99                                               >>> bbg.batch_size
   100                                               4
   101                                               >>> bbg.batch_num_nodes()
   102                                               tensor([4, 3, 4, 3])
   103                                               >>> bbg.batch_num_edges()
   104                                               tensor([3, 4, 3, 4])
   105                                           
   106                                               Batch graphs with feature data
   107                                           
   108                                               >>> g1.ndata['x'] = th.zeros(g1.num_nodes(), 3)
   109                                               >>> g1.edata['w'] = th.ones(g1.num_edges(), 2)
   110                                               >>> g2.ndata['x'] = th.ones(g2.num_nodes(), 3)
   111                                               >>> g2.edata['w'] = th.zeros(g2.num_edges(), 2)
   112                                               >>> bg = dgl.batch([g1, g2])
   113                                               >>> bg.ndata['x']
   114                                               tensor([[0, 0, 0],
   115                                                       [0, 0, 0],
   116                                                       [0, 0, 0],
   117                                                       [0, 0, 0],
   118                                                       [1, 1, 1],
   119                                                       [1, 1, 1],
   120                                                       [1, 1, 1]])
   121                                               >>> bg.edata['w']
   122                                               tensor([[1, 1],
   123                                                       [1, 1],
   124                                                       [1, 1],
   125                                                       [0, 0],
   126                                                       [0, 0],
   127                                                       [0, 0],
   128                                                       [0, 0]])
   129                                           
   130                                               Batch heterographs
   131                                           
   132                                               >>> hg1 = dgl.heterograph({
   133                                               ...     ('user', 'plays', 'game') : (th.tensor([0, 1]), th.tensor([0, 0]))})
   134                                               >>> hg2 = dgl.heterograph({
   135                                               ...     ('user', 'plays', 'game') : (th.tensor([0, 0, 0]), th.tensor([1, 0, 2]))})
   136                                               >>> bhg = dgl.batch([hg1, hg2])
   137                                               >>> bhg
   138                                               Graph(num_nodes={'user': 3, 'game': 4},
   139                                                     num_edges={('user', 'plays', 'game'): 5},
   140                                                     metagraph=[('drug', 'game')])
   141                                               >>> bhg.batch_size
   142                                               2
   143                                               >>> bhg.batch_num_nodes()
   144                                               {'user' : tensor([2, 1]), 'game' : tensor([1, 3])}
   145                                               >>> bhg.batch_num_edges()
   146                                               {('user', 'plays', 'game') : tensor([2, 3])}
   147                                           
   148                                               See Also
   149                                               --------
   150                                               unbatch
   151                                               """
   152        20         40.0      2.0      0.0      if len(graphs) == 0:
   153                                                   raise DGLError('The input list of graphs cannot be empty.')
   154        20         36.0      1.8      0.0      if node_attrs is not None:
   155                                                   dgl_warning('Arguments node_attrs has been deprecated. Please use'
   156                                                               ' ndata instead.')
   157                                                   ndata = node_attrs
   158        20         27.0      1.4      0.0      if edge_attrs is not None:
   159                                                   dgl_warning('Arguments edge_attrs has been deprecated. Please use'
   160                                                               ' edata instead.')
   161                                                   edata = edge_attrs
   162        20         48.0      2.4      0.0      if not (is_all(ndata) or isinstance(ndata, list)):
   163                                                   raise DGLError('Invalid argument ndata: must be a string list but got {}.'.format(
   164                                                       type(ndata)))
   165        20         37.0      1.9      0.0      if not (is_all(edata) or isinstance(edata, list)):
   166                                                   raise DGLError('Invalid argument edata: must be a string list but got {}.'.format(
   167                                                       type(edata)))
   168        20       4461.0    223.1      0.2      if any(g.is_block for g in graphs):
   169                                                   raise DGLError("Batching a block is not supported.")
   170                                           
   171        20     277936.0  13896.8     12.8      utils.check_all_same_device(graphs, 'graphs')
   172        20     255980.0  12799.0     11.7      utils.check_all_same_idtype(graphs, 'graphs')
   173        20        117.0      5.8      0.0      relations = list(sorted(graphs[0].canonical_etypes))
   174        20         68.0      3.4      0.0      ntypes = list(sorted(graphs[0].ntypes))
   175        20         61.0      3.0      0.0      etypes = [etype for _, etype, _ in relations]
   176        20        338.0     16.9      0.0      idtype = graphs[0].idtype
   177        20        435.0     21.8      0.0      device = graphs[0].device
   178                                           
   179        20     286362.0  14318.1     13.1      gidx = disjoint_union(graphs[0]._graph.metagraph, [g._graph for g in graphs])
   180        20       4933.0    246.7      0.2      retg = DGLHeteroGraph(gidx, ntypes, etypes).to(device)
   181                                           
   182                                               # Compute batch num nodes
   183        20         31.0      1.6      0.0      bnn = {}
   184        40         83.0      2.1      0.0      for ntype in ntypes:
   185        20      90601.0   4530.1      4.2          bnn[ntype] = F.cat([g.batch_num_nodes(ntype) for g in graphs], 0)
   186        20        127.0      6.3      0.0      retg.set_batch_num_nodes(bnn)
   187                                           
   188                                               # Compute batch num edges
   189        20         28.0      1.4      0.0      bne = {}
   190        40         75.0      1.9      0.0      for etype in relations:
   191        20      95621.0   4781.1      4.4          bne[etype] = F.cat([g.batch_num_edges(etype) for g in graphs], 0)
   192        20         95.0      4.8      0.0      retg.set_batch_num_edges(bne)
   193                                           
   194                                               # Batch node feature
   195        20         31.0      1.6      0.0      if ndata is not None:
   196        40         69.0      1.7      0.0          for ntype in ntypes:
   197        20     537658.0  26882.9     24.7              feat_dicts = [g.nodes[ntype].data for g in graphs if g.number_of_nodes(ntype) > 0]
   198        20      34001.0   1700.0      1.6              ret_feat = _batch_feat_dicts(feat_dicts, ndata, 'nodes["{}"].data'.format(ntype))
   199        20        396.0     19.8      0.0              retg.nodes[ntype].data.update(ret_feat)
   200                                           
   201                                               # Batch edge feature
   202        20         28.0      1.4      0.0      if edata is not None:
   203        40         67.0      1.7      0.0          for etype in relations:
   204        20     554844.0  27742.2     25.5              feat_dicts = [g.edges[etype].data for g in graphs if g.number_of_edges(etype) > 0]
   205        20      33508.0   1675.4      1.5              ret_feat = _batch_feat_dicts(feat_dicts, edata, 'edges[{}].data'.format(etype))
   206        20        405.0     20.2      0.0              retg.edges[etype].data.update(ret_feat)
   207                                           
   208        20         26.0      1.3      0.0      return retg

Total time: 0.10543 s
File: /home/expye/repo/dgl/python/dgl/heterograph.py
Function: _init at line 80

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
    80                                               @profile
    81                                               def _init(self, gidx, ntypes, etypes, node_frames, edge_frames):
    82                                                   """Init internal states."""
    83      1020       2339.0      2.3      2.2          self._graph = gidx
    84      1020       1512.0      1.5      1.4          self._canonical_etypes = None
    85      1020       1311.0      1.3      1.2          self._batch_num_nodes = None
    86      1020       1266.0      1.2      1.2          self._batch_num_edges = None
    87                                           
    88                                                   # Handle node types
    89      1020       1687.0      1.7      1.6          if isinstance(ntypes, tuple):
    90                                                       if len(ntypes) != 2:
    91                                                           errmsg = 'Invalid input. Expect a pair (srctypes, dsttypes) but got {}'.format(
    92                                                               ntypes)
    93                                                           raise TypeError(errmsg)
    94                                                       if not is_unibipartite(self._graph.metagraph):
    95                                                           raise ValueError('Invalid input. The metagraph must be a uni-directional'
    96                                                                            ' bipartite graph.')
    97                                                       self._ntypes = ntypes[0] + ntypes[1]
    98                                                       self._srctypes_invmap = {t : i for i, t in enumerate(ntypes[0])}
    99                                                       self._dsttypes_invmap = {t : i + len(ntypes[0]) for i, t in enumerate(ntypes[1])}
   100                                                       self._is_unibipartite = True
   101                                                       if len(ntypes[0]) == 1 and len(ntypes[1]) == 1 and len(etypes) == 1:
   102                                                           self._canonical_etypes = [(ntypes[0][0], etypes[0], ntypes[1][0])]
   103                                                   else:
   104      1020       1314.0      1.3      1.2              self._ntypes = ntypes
   105      1020       1505.0      1.5      1.4              if len(ntypes) == 1:
   106      1020       1327.0      1.3      1.3                  src_dst_map = None
   107                                                       else:
   108                                                           src_dst_map = find_src_dst_ntypes(self._ntypes, self._graph.metagraph)
   109      1020       1373.0      1.3      1.3              self._is_unibipartite = (src_dst_map is not None)
   110      1020       1314.0      1.3      1.2              if self._is_unibipartite:
   111                                                           self._srctypes_invmap, self._dsttypes_invmap = src_dst_map
   112                                                       else:
   113      1020       3173.0      3.1      3.0                  self._srctypes_invmap = {t : i for i, t in enumerate(self._ntypes)}
   114      1020       1267.0      1.2      1.2                  self._dsttypes_invmap = self._srctypes_invmap
   115                                           
   116                                                   # Handle edge types
   117      1020       1252.0      1.2      1.2          self._etypes = etypes
   118      1020       1308.0      1.3      1.2          if self._canonical_etypes is None:
   119      1020       1484.0      1.5      1.4              if (len(etypes) == 1 and len(ntypes) == 1):
   120      1020       1800.0      1.8      1.7                  self._canonical_etypes = [(ntypes[0], etypes[0], ntypes[0])]
   121                                                       else:
   122                                                           self._canonical_etypes = make_canonical_etypes(
   123                                                               self._etypes, self._ntypes, self._graph.metagraph)
   124                                           
   125                                                   # An internal map from etype to canonical etype tuple.
   126                                                   # If two etypes have the same name, an empty tuple is stored instead to indicate
   127                                                   # ambiguity.
   128      1020       1388.0      1.4      1.3          self._etype2canonical = {}
   129      2040       2941.0      1.4      2.8          for i, ety in enumerate(self._etypes):
   130      1020       1499.0      1.5      1.4              if ety in self._etype2canonical:
   131                                                           self._etype2canonical[ety] = tuple()
   132                                                       else:
   133      1020       1619.0      1.6      1.5                  self._etype2canonical[ety] = self._canonical_etypes[i]
   134      1020       2447.0      2.4      2.3          self._etypes_invmap = {t : i for i, t in enumerate(self._canonical_etypes)}
   135                                           
   136                                                   # node and edge frame
   137      1020       1280.0      1.3      1.2          if node_frames is None:
   138      1020       1654.0      1.6      1.6              node_frames = [None] * len(self._ntypes)
   139      1020       1350.0      1.3      1.3          node_frames = [Frame(num_rows=self._graph.number_of_nodes(i))
   140                                                                  if frame is None else frame
   141      1020      32363.0     31.7     30.7                         for i, frame in enumerate(node_frames)]
   142      1020       1503.0      1.5      1.4          self._node_frames = node_frames
   143                                           
   144      1020       1263.0      1.2      1.2          if edge_frames is None:
   145      1020       1688.0      1.7      1.6              edge_frames = [None] * len(self._etypes)
   146      1020       1357.0      1.3      1.3          edge_frames = [Frame(num_rows=self._graph.number_of_edges(i))
   147                                                                  if frame is None else frame
   148      1020      26140.0     25.6     24.8                         for i, frame in enumerate(edge_frames)]
   149      1020       1706.0      1.7      1.6          self._edge_frames = edge_frames
```

**Comment by yzh119** [comment]
Conclusion, when batching large number of small graphs, the time spent on `check_all_same_device` and `check_all_same_idtype` is non-trivial, we can move the check to C++ end.

The time spent on batching *emtpy* node/edge features is also non-negligible. @BarclayII what's your idea?
